### PR TITLE
[v2] Adds ConfigSync + Policy Controller

### LINKED ISSUE #50: Add Config Sync and Policy Controller
(Config Sync, Policy Controller, Config Connector)

- [x] Add CLI config fields (bools) to toggle those 3 things separately 
- [x] Write those bools to tfvars from the CLI 
- [x] Make sure that GKE cluster create (clusters.tf) uses the Config Connector bool as a flag 
- [x] if Config Sync enabled, use Terraform to create a Cloud Source Repository in the user's GCP project
- [x] post cluster create, use Terraform / googlebeta provider to register GKE clusters to Anthos (hub membership)
- [x] if Config Sync or Policy controller enabled, use TF enable the ACM feature -- this is project-wide
- [x] if Config sync or Policy Controller enabled, use TF to install ACM 
- [ ] [CLI -> INVOKE SCRIPT] initialize config sync repo with a directory structure + namespace 
- [ ] [CLI -> INVOKE SCRIPT] finish config connector install (`configconnector.yaml` and annotated `gcp` namespace)

### PR DESCRIPTION
**Pull request by askmeegs**
Partially fixes #50.

Adds support for ACM (Config Sync, Policy Controller) via golang and Terraform. 

Screen Shot 2021-11-08 at 5 25 08 AM

### How it works: 
- User sets values of bool vars `ConfigSync` and `PolicyController` in their `config.yaml` (both defaulted to true) 
- CLI writes those values to `tfvars`
- Terraform creates core ACM resources on apply (Cloud Source Repo, installs Config Sync and Policy Controller) 
- After Terraform apply, the CLI invokes `acm/configsync.go` which does the following post-install steps:
- Creates an SSH keypair for auth between Config Sync and Cloud Source Repos (`crypto` package)
- Prompts user to register the SSH public key in Cloud Source Repos (Cloud Console)
- If clusters have private endpoint, prompts user to open another terminal tab and run the equivalent of `make start proxy` (`gcloud ssh` to bastion host w/ tunneling) 
- `configsync.go` continues by connecting to all GKE clusters, generating an in-memory kubeconfig. (`client-go` package). If private clusters, client-go uses a `Proxy` for `localhost:8888`. This is the equivalent to setting the `HTTPS_PROXY` env var. 
- `configsync.go` creates the `gitcreds` secret for Config Sync to read from Cloud Source Repos. this secret is the value of the ssh private key. 
- CLI Prompts the user with the `git clone` command for their cloud source repo. 

##  ⚠️ Other changes in this PR 
- For public endpoint clusters, VPCConfig's `Auth IP` is now `Auth CIDR`, and defaults to wide internet access - `0.0.0.0/0`. 

## Notes 
- There is no script invocation here, just Golang. **But** the user is prompted 3 times to open a browser/run a command. (ssh register, start proxy, clone repo). ssh register seems impossible to do with a client lib/command; the golang compute engine client lib doesn't expose an `ssh` function (needed for proxy), and the golang git clients I tried to use gave me trouble in SSH land. The last one (git clone, git push) we *could* probably do in go and this would automate those final Config Sync setup steps. 
- All GKE clusters use the same repo for sync. 
- Config Sync uses an unstructured, not hierarchical repo. that's hardcoded right now. 
- Until the user clones their repo and pushes a YAML config, config sync will error because the `main` branch isn't initialized yet. 
- I reversed the private endpoint logic, it seemed wrong to me 

## TODO 
- (Another PR) Add Config Connector support and post-install. This should be easy-ish now that we can use `client-go` to apply K8s resources after terraform create.

### CODE REVIEW

**Comment by askmeegs** [comment]
@googlebot I fixed it.

File: cli/pkg/acm/configsync.go [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+	// Basic config structure
+	ret := api.Config{
+		APIVersion: "v1",
+		Kind:       "Config",
+		Clusters:   map[string]*api.Cluster{},  // Clusters is a map of referencable names to cluster configs
+		AuthInfos:  map[string]*api.AuthInfo{}, // AuthInfos is a map of referencable names to user configs
+		Contexts:   map[string]*api.Context{},  // Contexts is a map of referencable names to context configs
+	}
+
+	// Ask Google for a list of all kube clusters in the given project.
```
Comment (knee-berts) [inline cli/pkg/acm/configsync.go:169]: Note for myself. We need to call out in the docs that projects used with the toolkit should be empty.

**Review by knee-berts — APPROVED** [review]
LGTM - Excellent work Megan, well done!

File: cli/samples/multi-cluster.yaml [hunk 2]
```diff
@@ -15,7 +15,7 @@ vpcConfig:
   vpcProjectId: "my-project" # Required if you are using a sharedVPC
   podCIDRName: "mypodcidr"
   svcCIDRName: "mysvccidr"
-  authIp: "10.10.10.10" # only needed if enabling private endpoints for clusters
+  authCIDR: "0.0.0.0/0" # only needed if enabling private endpoints for clusters
```
Comment (askmeegs) [inline cli/samples/multi-cluster.yaml:18]: wait @knee-berts shouldn't the comment say `public` endpoints, not `private` ?

File: terraform/cluster_build/main.tf [hunk 3]
```diff
@@ -198,3 +198,15 @@ module "kms" {
     "serviceAccount:${local.clu_service_account}",
   ]
 }
+
+module "acm" {
+  depends_on = [
+    module.gke,
+  ]
+  count             = var.config_sync ? 1 : 0
+  source            = "./acm/"
```
Comment (albertwo1978) [inline terraform/cluster_build/main.tf:207]: could we rename the folder to /modules/ ? It's a better pattern in the chance we need to add more modules in the future.

Comment (knee-berts) [reply terraform/cluster_build/main.tf:207]: Al, do you want the ACM folder renamed to modules, or nested in a new modules folder? I think the latter would be best for future local modules. "clusterbuild/modules/ACM".

**Review by knee-berts — CHANGES_REQUESTED** [review]
LGTM aside from Al's suggestion to add a modules folder for the ACM local module to live in.

**Pull request merged**